### PR TITLE
[Merged by Bors] - Enable `-race` flag in unit tests

### LINKED ISSUE #2226: Add -race flag to tests and fix data races
## Description
Neither unit tests, nor system tests use the `-race` flag. Using it would enable Go's race detector and let us find some existing data races easily. As a result of a read/write operation is undefined in case of a data race, they might cause some unexpected behavior.

I ran unit tests with the flag, and there are many races in multiple packages.

This issue appears in commit hash: 
`f790a44`

Related files (optionally with line numbers): 
https://github.com/spacemeshos/go-spacemesh/blob/develop/Makefile#L137
https://github.com/spacemeshos/go-spacemesh/blob/develop/Makefile#L142
https://github.com/spacemeshos/go-spacemesh/blob/develop/Makefile#L142
https://github.com/spacemeshos/go-spacemesh/blob/develop/Dockerfile#L31

### PR DESCRIPTION
**Pull request by nkryuchkov**
## Motivation
Closes #2226
There're some data races left after merging https://github.com/spacemeshos/go-spacemesh/pull/2608. I cannot reproduce them locally but in CI only. Therefore, I'm creating this PR to be able to check what's left to be fixed.

## Changes

- Enable `-race` flag in unit tests

## Test Plan
- Unit tests

## DevOps Notes
- [x] This PR does not require configuration changes (e.g., environment variables, GitHub secrets, VM resources)
- [x] This PR does not affect public APIs
- [x] This PR does not rely on a new version of external services (PoET, elasticsearch, etc.)
- [x] This PR does not make changes to log messages (which monitoring infrastructure may rely on)

### CODE REVIEW

**Comment by nkryuchkov** [comment]
On hold until https://github.com/spacemeshos/go-spacemesh/issues/3066, https://github.com/spacemeshos/go-spacemesh/issues/3067, https://github.com/spacemeshos/go-spacemesh/issues/3097, https://github.com/spacemeshos/go-spacemesh/issues/3099, https://github.com/spacemeshos/go-spacemesh/issues/3100 are closed.

**Comment by nkryuchkov** [comment]
bors try

**Comment by dshulyak** [comment]
do we need to merge any other changes?

**Comment by nkryuchkov** [comment]
@dshulyak we need to fix https://github.com/spacemeshos/go-spacemesh/issues/3100, to check if https://github.com/spacemeshos/go-spacemesh/issues/3066 is reproducible and then to check if any other data races appear

**Comment by nkryuchkov** [comment]
@dshulyak https://github.com/spacemeshos/go-spacemesh/pull/3095 should fix a mesh data race and I think it's still not merged here

**Comment by poszu** [comment]
bors try

File: cmd/node/node.go
Comment (fasmat) [inline cmd/node/node.go:98]: I think if one calls `GetCommand()` mulitple times this might cause issues, since there is a side effect on line 138 with

```golang
	cmdp.AddCommands(cmd)
```

Can you wrap this into a `sync.Once` to make sure that multiple calls to `GetCommand()` always return the same instance (and still be concurrency safe?)

File: hare/hare_rounds_test.go
Comment (fasmat) [inline hare/hare_rounds_test.go:127]: You can add `defer cancel()` here. Calling cancel after the first time is a no-op. Without the defer, if the test fails between here and line 147 we have a dead go routine.

**Review by fasmat — CHANGES_REQUESTED** [review]
We should take a look at the side-effect of `GetCommand()` besides that the PR lgtm

File: cmd/node/node.go
Comment (poszu) [reply cmd/node/node.go:98]: What is the side effect? This call adds flags to local `cmd`.

Comment (fasmat) [reply cmd/node/node.go:98]: You're right; I was confused by the name `AddCommands` and assumed that `cmd` would be added to something. A better name for the function would probably be `ParseCmdFlags` or similar. But yes `sync.Once` isn't needed here.

**Review by fasmat — APPROVED** [review]

**Review by countvonzero — APPROVED** [review]
yay!

**Comment by fasmat** [comment]
I hate being a buzz-kill, but there are still some races left 🙁. Check the Windows unittests. At least some of them fail because of data races. The macOS tests seem to fail because of timing issues at least.

Also `libgpu-setup.dynlib` seems to have been built on a newer version of macOS than we are currently using in our `go-spacemesh` CI. I should double check our CI for `go-spacemesh` and `gpu-post`, so that we can support at all macOS versions since at least 11.0 (Big Sur).

**Comment by poszu** [comment]
> I hate being a buzz-kill, but there are still some races left 🙁. Check the Windows unittests. At least some of them fail because of data races. The macOS tests seem to fail because of timing issues at least.
> 
> Also `libgpu-setup.dynlib` seems to have been built on a newer version of macOS than we are currently using in our `go-spacemesh` CI. I should double check our CI for `go-spacemesh` and `gpu-post`, so that we can support at all macOS versions since at least 11.0 (Big Sur).

I'm aware there are still data races in the code. The goal of this PR is to only fix failing tests and to enable `-race` so that no new data races are merged. 

Regarding windows unit tests - I'm surprised as I was pretty sure I removed the `allow-failure: true` for windows some time ago 🤔 .

**Comment by poszu** [comment]
bors merge

**Pull request closed without merging**
